src/biokbase/narrative/tests/test_upa_api.py

```diff
@@ -0,0 +1,116 @@
+"""
+Tests for the Python side of the UPA api.
+"""
+import unittest
+import mock
+from biokbase.narrative.upa import (
+    external_tag,
+    serialize,
+    deserialize
+)
+import os
+
+this_ws = "31"
+
+
+def mock_sys_var(var):
+    return this_ws
+
+
+class UpaApiTestCase(unittest.TestCase):
+    @classmethod
+    def setUpClass(self):
+        self.test_data = [{
+            "upa": "31/2/3",
+            "serial": "[31]/2/3"
+        }, {
+            "upa": "31/2/3;4/5/6",
+            "serial": "[31]/2/3;4/5/6"
+        }, {
+            "upa": "31/2/3;4/5/6;7/8/9",
+            "serial": "[31]/2/3;4/5/6;7/8/9"
+        }, {
+            "upa": ["31/2/3", "4/5/6", "7/8/9"],
+            "serial": "[31]/2/3;4/5/6;7/8/9"
+        }, {
+            "upa": "31/31/31;31/31/31",
+            "serial": "[31]/31/31;31/31/31"
+        }, {
+            "upa": "5/1/2",
+            "serial": "{}5/1/2".format(external_tag)
+        }, {
+            "upa": "5/1/2;1/2/3",
+            "serial": "{}5/1/2;1/2/3".format(external_tag)
+        }, {
+            "upa": "5/5/5;1/1/1",
+            "serial": "{}5/5/5;1/1/1".format(external_tag)
+        }]
+
+        self.bad_upas = [
+            "1/2",
+            "1/2/3;4/5/",
+            "1/2/a",
+            "123/456/7/",
+            "1/2/3;",
+            "x/y/z"
```

Comment:
Try some upas with ws & obj names, or just names alone, or just numbers alone, with and without refpaths.

Reply:
done